File: app/src/main/java/com/example/potato/shakyroads/MainActivity.java
package com.example.potato.shakyroads;

import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Environment;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.util.Log;
import android.view.View;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.widget.TextView;
import android.Manifest;
import com.opencsv.CSVWriter;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.OutputStreamWriter;
import java.lang.reflect.Array;
import java.nio.charset.StandardCharsets;

import static java.lang.String.valueOf;


public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener, SensorEventListener, LocationListener {


    private SensorManager mSensorManager;
    private LocationManager mLocationManager;
    private Sensor mAccelerometer;
    private Context mContext;
    public static final int MY_PERMISSIONS_REQUEST_LOCATION = 99;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "This button does nothing", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        // get instances of SensorManager
        mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);

        // get and instance of the linear motion accelerometer
        mAccelerometer = mSensorManager.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION);

        // get an instance of LocationManager
        mLocationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
    }


    /* when the activity resumes */
    @Override
    protected void onResume() {
        super.onResume();

        // Start the reading from the accelerometer
        startAccelerometer();
        // start getting location updates
        startLocation();
    }

    /* when the activity pauses */
    @Override
    protected void onPause() {
        super.onPause();

        // When the activity is paused, we make sure to release our sensor resources
        stopAccelerometer();
        // stop the location listeners
        stopLocation();
    }


    /*
    Display functions for testing.
     */
    /* displays the reading */
    public void displayAcceleration(double acceleration) {
        TextView display = (TextView)findViewById(R.id.reading);
        display.setText(valueOf(acceleration));
    }

    /* displays the longitude */
    public void displayLong(double location) {
        TextView display = (TextView)findViewById(R.id.longitude);
        display.setText(valueOf(location));
    }

    /* displays the latitude */
    public void displayLat(double location) {
        TextView display = (TextView)findViewById(R.id.latitude);
        display.setText(valueOf(location));
    }

    /* gets the gps permission if we don't have it, if it has the permission: start the GPS */
    public void startGPS(View view) {
        checkLocationPermission();
        startLocation();
    }


    /* Checks if external storage is available for read and write */
    public boolean isExternalStorageWritable() {
        String state = Environment.getExternalStorageState();
        if (Environment.MEDIA_MOUNTED.equals(state)) {
            return true;
        }
        return false;
    }

    /* Writes the readings to a csv file */
    public void saveData(double lat, double lng, double acc) {
        if (isExternalStorageWritable() == true) {
            try {
                // check if the csv file exists, and make one if it doesn't
                String fileName = "shakyroads.csv";
                File file = new File(Environment.getExternalStoragePublicDirectory(
                        Environment.DIRECTORY_DOCUMENTS), fileName);
                if(!file.exists()) {
                    file = new File(Environment.getExternalStoragePublicDirectory(
                            Environment.DIRECTORY_DOCUMENTS), fileName);
                }
                if (!file.mkdirs()) {
                    Log.e("saveData", "Directory not created");
                    file.mkdirs();
                    // TODO make the directory instead of just complaining.
                }

                CSVWriter writer = new CSVWriter(
                        new OutputStreamWriter(new FileOutputStream(fileName),
                                StandardCharsets.UTF_8), ',', '"', '"', "\n");

                // convert the data to strings
                String latitude = String.valueOf(lat);
                String longitude = String.valueOf(lng);
                String acceleration = String.valueOf(acc);
                // then to an array of strings
                String[] entries = new String[] {latitude, longitude, acceleration};

                // then write it
                writer.writeNext(entries);
                writer.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }



    ////////////////////////////////////////////////////////////////////////////////////////////////
    // The accelerometer stuff
    ////////////////////////////////////////////////////////////////////////////////////////////////

    /* start the sensor listener */
    public void startAccelerometer() {
        mSensorManager.registerListener(this, mAccelerometer, SensorManager.SENSOR_DELAY_NORMAL);
    }

    /* stop the sensor listener */
    public void stopAccelerometer() {
        mSensorManager.unregisterListener(this);
    }

    /* called whenever the accelerometer picks up any movement */
    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType()==Sensor.TYPE_LINEAR_ACCELERATION) {
            double x = event.values[0];
            double y = event.values[1];
            double z = event.values[2];

            displayAcceleration(y);

            saveData(0, 0, y);
        }
    }

    // leave this here, it's not used
    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {}


    ////////////////////////////////////////////////////////////////////////////////////////////////
    // The GPS stuff
    ////////////////////////////////////////////////////////////////////////////////////////////////

    /*
    This checks if the app has the permissions to use the GPS, if the app doesn't have
    the permissions then it shows the user a dialog asking for the permission.
    Returns true or false depending on whether it has the permission or not
     */
    public boolean checkLocationPermission() {
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.ACCESS_FINE_LOCATION)) {

                // Show an explanation to the user *asynchronously* -- don't block
                // this thread waiting for the user's response! After the user
                // sees the explanation, try again to request the permission.
                new AlertDialog.Builder(this)
                        .setTitle(R.string.title_location_permission)
                        .setMessage(R.string.text_location_permission)
                        .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                //Prompt the user once explanation has been shown, used to be MainActivity.this
                                ActivityCompat.requestPermissions(MainActivity.this,
                                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                                        MY_PERMISSIONS_REQUEST_LOCATION);
                            }
                        })
                        .create()
                        .show();

            } else {
                // No explanation needed, we can request the permission.
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                        MY_PERMISSIONS_REQUEST_LOCATION);
            }
            return false;
        } else {
            return true;
        }
    }

    /*
    Checks if the app got permission to use the GPS, if it did get the permission then it
    calls startLocation(). startLocation() has a check as well to see if the app has the permission
    to use the GPS.
     */
    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           String permissions[], int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_LOCATION: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    // permission was granted, do the location-related task you need to do.
                    if (ContextCompat.checkSelfPermission(this,
                            Manifest.permission.ACCESS_FINE_LOCATION)
                            == PackageManager.PERMISSION_GRANTED) {

                        //Request location updates:
                        startLocation();
                    }

                } else {

                    // TODO: disable the thing that depends on this permission.
                }
                return;
            }
        }
    }


    /* Register the listener with the Location Manager to receive location updates */
    public void startLocation() {
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED) {

            mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, this);
        }
    }

    /* Remove the location listener to stop receiving location updates */
    public void stopLocation() {
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED) {

            mLocationManager.removeUpdates(this);
        }

    }

    /* Called when location changes */
    @Override
    public void onLocationChanged(Location location) {
        double latitude = (double) (location.getLatitude());
        double longitude = (double) (location.getLongitude());

        displayLong(longitude);
        displayLat(latitude);
    }

    // these are not used, leave them here
    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {}

    @Override
    public void onProviderEnabled(String provider) {}

    @Override
    public void onProviderDisabled(String provider) {}



    ////////////////////////////////////////////////////////////////////////////////////////////////
    //The UI functions and handlers
    ////////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        if (id == R.id.nav_camera) {
            // Handle the camera action
        } else if (id == R.id.nav_gallery) {

        } else if (id == R.id.nav_slideshow) {

        } else if (id == R.id.nav_manage) {

        } else if (id == R.id.nav_share) {

        } else if (id == R.id.nav_send) {

        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

}

File: app/src/main/java/com/example/potato/shakyroads/MainActivity.kt
// this code is terrible please don't laugh
package com.example.potato.shakyroads

import android.content.Context
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Environment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AlertDialog
import android.util.Log
import android.view.View
import com.google.android.material.navigation.NavigationView
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.hardware.Sensor
import android.hardware.SensorManager
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.widget.TextView
import android.Manifest
import android.content.Intent
import android.widget.Toast
import androidx.preference.PreferenceManager

import com.opencsv.CSVWriter

import java.io.File
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets
import java.util.Calendar


class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener,
        SensorEventListener, LocationListener {

    private var mSensorManager: SensorManager? = null
    private var mAccelerometer: Sensor? = null
    private var mLocationManager: LocationManager? = null
    private val myPermissionRequestLocation = 99
    // vars to hold the readings
    private var globX: Double = 0.0
    private var globY: Double = 0.0
    private var globZ: Double = 0.0
    // stores the state of the button that toggles location
    private var isButtonPressed: Int = 0 // 0 = off, 1 = on
    private var dateTime = ""
    private var deltaTime = mutableListOf<Long>() // used to calc time passed


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        /*Button to toggle the logger*/
        val fab = findViewById<View>(R.id.fab) as FloatingActionButton
        fab.setOnClickListener { view ->
            isButtonPressed = 1 - isButtonPressed // math tricks: 1 - 0 = 1 and 1 - 1 = 0.
            // TODO: change the button icon rather than showing a SnackBar
            Snackbar.make(view, isButtonPressed.toString(), Snackbar.LENGTH_SHORT)
                    .setAction("Action", null).show()
            checkLocationPermission()
            // for time-stamping files, worst possible way to do it but hey im dumb
            dateTime = Calendar.getInstance().time.toString()
        }

        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        val toggle = ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        val navigationView = findViewById<View>(R.id.nav_view) as NavigationView
        navigationView.setNavigationItemSelectedListener(this)

        // get an instance of SensorManager & linear motion accelerometer
        mSensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
        mAccelerometer = mSensorManager!!.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)

        // get an instance of LocationManager
        mLocationManager = getSystemService(Context.LOCATION_SERVICE) as LocationManager
    }


    override fun onResume() {
        super.onResume()
        // resume reading from the sensors
        mSensorManager!!.registerListener(this, mAccelerometer, SensorManager.SENSOR_DELAY_FASTEST)
        startLocation()
    }

    override fun onStop() {
        super.onStop()
        // stop reading from the sensors
        mSensorManager!!.unregisterListener(this)
        stopLocation()
    }

    /* used to display stuff */
    private fun display(data: Double, viewId: Int) {
        val display = findViewById<View>(viewId) as TextView
        display.text = data.toString()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    // stuff that saves the data
    ////////////////////////////////////////////////////////////////////////////////////////////////


    /* Checks if external storage is available to read and write */
    private fun isExternalStorageWritable(): Boolean {
        return Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED
    }

    /* writes the GPS and acceleration data to a csv file */
    private fun saveData(accX: Double, accY: Double, accZ: Double, lat: Double, lng: Double) {
        // filename and path to file
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        val fileName = sharedPreferences.getString("filename", "ShakyroadsData")
        val path = Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_DOCUMENTS + "/shakyroads/")
        val file = File(path, "$fileName $dateTime.csv")


        if (isExternalStorageWritable()) {
            try {
                // make sure the directory exists
                path.mkdirs()

                // get an instance of CSVWriter
                val writer = CSVWriter(
                        OutputStreamWriter(FileOutputStream(file, true),
                                StandardCharsets.UTF_8), ',', CSVWriter.NO_QUOTE_CHARACTER, "\n")

                // write em
                val entries = arrayOf(lat.toString(), lng.toString(), accX.toString(),
                        accY.toString(), accZ.toString())
                writer.writeNext(entries)
                writer.close()

            } catch (e: Exception) {
                e.printStackTrace()
                Log.e("holy crap", "FileOutputStream threw an IOException")
            }
        } else {
            Log.e("saveFile", "storage not writable, you probably don't have the privs")
            // TODO ask for permission instead of just complaining
        }
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    // The accelerometer stuff
    ////////////////////////////////////////////////////////////////////////////////////////////////

    /* called whenever the accelerometer picks up any movement */
    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type == Sensor.TYPE_LINEAR_ACCELERATION) {
            globX = event.values[0].toDouble()
            globY = event.values[1].toDouble()
            globZ = event.values[2].toDouble()

            // calculate "jerk", m*s^-3
            deltaTime.add(System.currentTimeMillis())

            val elapsedTime: Long = if (deltaTime.size >= 3) {
                deltaTime.removeAt(0) // pop off old readings
                deltaTime[0] - deltaTime[1]
            } else {
                0
            }
            // please forgive me
            globX /= elapsedTime; globY /= elapsedTime; globZ /= elapsedTime
            display(globX, R.id.X)
            display(globY, R.id.Y)
            display(globZ, R.id.Z)
            if (isButtonPressed == 1) {
                saveData(globX, globY, globZ, 0.0, 0.0)
            }
        }
    }

    // leave this here, it's not used
    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}


    ////////////////////////////////////////////////////////////////////////////////////////////////
    // The GPS stuff
    ////////////////////////////////////////////////////////////////////////////////////////////////

    /*
    This checks if the app has the permissions to use the GPS, if the app doesn't have
    the permissions then it shows the user a dialog asking for the permission.
    Returns true or false depending on whether it has the permission or not
     */
    private fun checkLocationPermission(): Boolean {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                            Manifest.permission.ACCESS_FINE_LOCATION)) {

                // Show an explanation to the user asynchronously. After the user
                // sees the explanation, try again to request the permission.
                AlertDialog.Builder(this)
                        .setTitle(R.string.title_location_permission)
                        .setMessage(R.string.text_location_permission)
                        .setPositiveButton(R.string.ok) { _, _ ->
                            //Prompt the user once explanation has been shown
                            ActivityCompat.requestPermissions(this@MainActivity,
                                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                                    myPermissionRequestLocation)
                        }
                        .create()
                        .show()
            } else {
                // No explanation needed, we can request the permission.
                ActivityCompat.requestPermissions(this,
                        arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                        myPermissionRequestLocation)
            }
            return false
        } else {
            return true
        }
    }

    /*
    Checks if the app got permission to use the GPS, if it did get the permission then it
    calls startLocation(). startLocation() has a check as well to see if the app has the permission
    to use the GPS.
     */
    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            myPermissionRequestLocation -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    // permission was granted, do the location-related task you need to do.
                    if (ContextCompat.checkSelfPermission(this,
                                    Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                        startLocation()
                    }
                } else {
                    // TODO: make it obvious the app can't do its thing
                    stopLocation()
                }
                return
            }
        }
    }


    /* Register the listener with the Location Manager to receive location updates */
    private fun startLocation() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED) {
            // minTime: milliseconds, minDistance: meters
            mLocationManager!!.requestLocationUpdates(LocationManager.GPS_PROVIDER,
                    (1 * 1000).toLong(), 2f, this)
        }
    }

    /* Remove the location listener to stop receiving location updates */
    private fun stopLocation() {
        mLocationManager!!.removeUpdates(this)
    }

    /* Called when location changes */
    override fun onLocationChanged(location: Location) {
        val latitude = location.latitude
        val longitude = location.longitude

        display(longitude, R.id.latitude)
        display(latitude, R.id.longitude)

//        if (isButtonPressed == 1) {
//            saveData(globX, globY, globZ, latitude, longitude)
//        }
    }

    // these are not used yet, leave them here
    override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}

    override fun onProviderEnabled(provider: String) {}

    override fun onProviderDisabled(provider: String) {
        // tell the user when the location provider is disabled
        stopLocation()
        Toast.makeText(this, "GPS is turned off", Toast.LENGTH_LONG).show()
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    //The UI functions and handlers
    ////////////////////////////////////////////////////////////////////////////////////////////////
    override fun onBackPressed() {
        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId

        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        when (item.itemId) {
            R.id.nav_camera -> {

            }
            R.id.nav_settings -> {
                val intent = Intent(this, SettingsActivity::class.java)
                startActivity(intent)
            }
            R.id.nav_share -> {

            }
            R.id.nav_send -> {

            }
        }
        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        drawer.closeDrawer(GravityCompat.START)
        return true
    }
}